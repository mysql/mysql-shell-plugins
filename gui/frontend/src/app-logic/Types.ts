/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

// TODO: check if we can switch to camel case for those DB data types that use snake case.

/* eslint-disable @typescript-eslint/naming-convention */

import * as codicon from "../components/ui/Codicon";

// Types used in different places in the application (modules, parser, scripting etc.).

export interface IDictionary {
    [key: string]: unknown;
}

// Types of columns in a table column definition, which we support in the application.
// Also used for column affinities in SQLite (instead of column types).
export enum DBDataType {
    Unknown,

    // Integer types.
    TinyInt,
    SmallInt,
    MediumInt,
    Int,                // Any integer like number type (numeric is essentially the same) in SQLite.
    Bigint,
    UInteger,           // Mapped type from the shell.

    // Float types.
    Float,
    Real,               // Any floating point type in SQLite.
    Double,
    Decimal,

    // Binary types.
    Binary,
    Varbinary,

    // Text types.
    Char,
    Nchar,
    Varchar,
    Nvarchar,

    String,             // Mapped type from the shell.
    TinyText,
    Text,               // Any text in SQLite.
    MediumText,
    LongText,

    // Blob types.
    TinyBlob,
    Blob,               // NULL or no (explicit) type in SQLite.
    MediumBlob,
    LongBlob,

    // Date and time types.
    DateTime,
    DateTime_f,
    Date,
    Time,
    Time_f,
    Year,
    Timestamp,
    Timestamp_f,

    // Geometry types.
    Geometry,
    Point,
    LineString,
    Polygon,
    GeometryCollection,
    MultiPoint,
    MultiLineString,
    MultiPolygon,

    // Special types.
    Numeric,            // SQLite only, similar to INTEGER.
    Json,
    Bit,
    Boolean,
    Enum,
    Set,
}

// Describes the number of parameters for a DB data type. It doesn't say anything about the parameter types, though.
export enum ParameterFormatType {
    None = 0,
    One = 1,            // (n)
    OneOrZero = 2,      // [(n)]
    Two = 3,            // (m,n)
    TwoOrOne = 4,       // (m[,n])
    TwoOrZero = 5,      // [(m,n)]
    TwoOrOneOrZero = 6, // [(m[, n])]
    List = 10,          // (m, n, o, p ...)
}

export interface IDBCharacterSet {
    collations: string[];
    defaultCollation: string;
    description: string;
}

// Describes a data type in a database.
export interface IDBDataTypeDetails {
    type: DBDataType;

    characterMaximumLength?: number;
    characterOctetLength?: number;
    dateTimePrecision?: number;
    flags?: string[];
    numericPrecision?: number;
    numericPrecisionRadix?: number;
    numericScale?: number;
    needsQuotes?: boolean;
    parameterFormatType?: ParameterFormatType;
    synonyms?: string[];
}

export interface IColumnInfo {
    /** The column's title shown in the UI. */
    title: string;

    /** The name of the field in the data, which contains the data for this column. */
    field: string;

    /** Note: SQLite has no column data types like other RDBMS-es. Instead we store column affinities here. */
    dataType: IDBDataTypeDetails;
}

export enum MessageType {
    Error,
    Warning,
    Info,
    Interactive,
    Response,
}

export interface IExecutionInfo {
    type?: MessageType;
    text: string;
}

// Types for requesting a specific dialog and sending back the entered values.

export enum DialogType {
    Prompt,     // A simple prompt value dialog, requesting a single value from the user.
    Confirm,    // Confirm a question (yes, no, alt).
    Select,     // Select one entry from a list.
    MrsService, // A dialog to create or edit an MRS service.
    MrsSchema,
}

/**
 * A set of values that describe a single modal dialog request.
 */
export interface IDialogRequest extends IDictionary {
    type: DialogType;

    /** An id to identify the invocation. */
    id: string;

    /** Optionally used to set a customized dialog title (where supported). */
    title?: string;

    /** A list of strings to be rendered before the actual prompt, each in an own paragraph. */
    description?: string[];

    /** Values to configure how the dialog looks like (available options in drop downs etc.). */
    parameters?: IDictionary;

    /** Values to pre-fill certain elements or additional data for captions, payload etc. */
    values?: IDictionary;

    /** Additional data for specific use cases. */
    data?: IDictionary;
}

/**
 * What decision made the user to close a dialog.
 */
export enum DialogResponseClosure {
    Accept,
    Decline,
    Alternative,

    /** Set when no decision was made. */
    Cancel,
}

export interface IDialogResponse extends IDictionary {
    id: string;
    type: DialogType;
    closure: DialogResponseClosure;

    values?: IDictionary;
    data?: IDictionary;
}

export interface IStatusbarInfo {
    id: string;
    text?: string;
    icon?: string | codicon.Codicon;
    choices?: Array<{ label: string; data: IDictionary }>;
    visible?: boolean;
    standout?: boolean;
}

/**
 * Defines the structure for password requests sent by various access methods. Passwords are stored by the backend
 * using native OS methods.
 */
export interface IServicePasswordRequest {
    /** The id of the request that asked for the password. */
    requestId: string;

    /** The password dialog title. */
    caption?: string;

    /** A list of strings describing the circumstances of the password request. */
    description?: string[];

    /** A human readable string describing the service. */
    service?: string;

    /** The user name for which to get the password. */
    user?: string;

    /** Any other data the caller wants to pass over. */
    payload?: IDictionary;
}
