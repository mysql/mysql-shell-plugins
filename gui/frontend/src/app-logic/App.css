/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

@viewport {
    width: device-width;
    zoom: 1.0;
}

@font-face {
    font-family: "SourceCodePro+Powerline+Awesome+MySQL Regular";
    font-style: normal;
    font-weight: 400;
    src: url(../assets/fonts/MySQL.ttf) format("truetype");
}

@font-face {
    font-family: "SourceCodePro+Powerline+Awesome+MySQL";
    font-style: normal;
    font-weight: 400;
    src: url(../assets/fonts/MySQL.ttf) format("truetype");
}

* {
    box-sizing: border-box;
}

html,
body {
    min-height: 100vh;
    margin: 0;
}

body {
    display: flex;
    flex-direction: column;
    margin: 0;
    padding: 0;
    overflow: hidden;

    color: var(--foreground);
    background-color: transparent;

    font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
    font-weight: 400;
    font-size: 15px;

    user-select: none;
}

#root {
    display: flex;
    flex-direction: column;
    height: 100vh;

    background-color: transparent;
}

a {
    color: var(--textLink-foreground);
}

a:active {
    color: var(--textLink-activeForeground);
}

pre {
    color: var(--textPreformat-foreground);
}

.verticalCenterContent {
    display: flex;
    flex-direction: row;
    align-items: center;
}

.centerContent {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
}

.ellipsis {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}

.msg .dropShadow {
    filter: drop-shadow(3px 3px 5px var(--widget-shadow));
}

.copyright::after {
    content: "Â© 2020, 2022, Oracle Corporation and/or its affiliates.";
    display: inline-block;
    margin-top: 80px;
    font-size: 10px;
    font-weight: 100;
    color: var(--text-dimmedForeground);
}

html.themeSwitch,
html.themeSwitch *,
html.themeSwitch *:before,
html.themeSwitch *:after {
    transition: color 500ms, background-color 500ms !important;
    transition-delay: 0 !important;
}

.msg .manualFocus,
.msg [tabindex="0"]:focus,
.msg input[type=button]:focus,
.msg input[type=text]:focus,
.msg input[type=password]:focus,
.msg input[type=number]:focus,

.msg button[search-type=option]:focus,
.msg.searchContainer:focus-within,

.msg select:focus,
.msg textarea:focus,
.msg button:focus {
    outline-width: 1px;
    outline-style: solid;
    outline-offset: -1px;
    outline-color: var(--focusBorder);
    opacity: 1 !important
}

/* The next two rules disable the number input spin buttons for all browsers. */
.msg input[type="number"] {
    -webkit-appearance: textfield;
    -moz-appearance: textfield;
    appearance: textfield;
}

.msg input[type=number]::-webkit-inner-spin-button,
.msg input[type=number]::-webkit-outer-spin-button {
    -webkit-appearance: none;
}

.msg::selection {
    background-color: var(--selection-background);
}

.msg.applicationHost {
    overflow: hidden;
    align-items: stretch;
    height: 100vh;

    background-color: var(--background);
}

.msg.applicationHost .moduleHost {
    display: none;
}

.msg.applicationHost .moduleHost.active {
    display: flex;
}

/* Special styles for macOS auto disappearing scroll bars. */
.msg .fixedScrollbar {
    overflow: auto;

    -webkit-overflow-scrolling: touch;

    scrollbar-width: auto;
    scrollbar-color: var(--scrollbarSlider-background) var(--scrollbarSlider-shadow);
}

.msg .fixedScrollbar::-webkit-scrollbar {
    -webkit-appearance: none;
    width: 7px;
    height: 7px;

    background-color: var(--scrollbarSlider-shadow);
}

.msg .fixedScrollbar::-webkit-scrollbar-thumb {
    border-radius: 4px;
    background-color: var(--scrollbarSlider-background);
}

.msg .fixedScrollbar::-webkit-scrollbar-corner {
    background-color: var(--scrollbarSlider-shadow);
}

/* Some browser specific rules. */

/* Removes the additional dotted focus indicator only visible in Firefox. */
::-moz-focus-inner {
    border: 0;
}
