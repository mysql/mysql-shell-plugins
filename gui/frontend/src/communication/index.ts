/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { IDictionary } from "../app-logic/Types";

export * from "./Protocol";
export * from "./ProtocolGui";
export * from "./ProtocolMds";
export * from "./ProtocolMrs";

export * from "./MessageScheduler";
export * from "./GeneralEvents";
export * from "./OciEvents";
export * from "./Oci";

// Types for communication in various environments, like a browser, native platform wrappers or Visual Studio Code.

// Determines the origin of the message.
export type IEmbeddedSourceType =
    "app" | // The app itself.
    "host"  // The host where the app is embedded.
    ;

export interface IEmbeddedMessage {
    source: IEmbeddedSourceType;
    command: string;

    data?: IDictionary;
}
