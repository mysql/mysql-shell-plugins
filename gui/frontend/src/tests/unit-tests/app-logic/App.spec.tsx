/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mount } from "enzyme";
import React from "react";

import { App } from "../../../app-logic/App";
import { IStatusbarInfo } from "../../../app-logic/Types";
import { MessageScheduler } from "../../../communication";
import { IEditorStatusInfo } from "../../../modules/db-editor";
import { appParameters, requisitions } from "../../../supplement/Requisitions";
import { waitFor } from "../../../utilities/helpers";
import { MySQLShellLauncher } from "../../../utilities/MySQLShellLauncher";
import { dispatchErrorResponse, setupShellForTests, snapshotFromWrapper } from "../test-helpers";
import { eventMock } from "../__mocks__/MockEvents";

describe("Application tests", () => {
    let app: ReturnType<typeof mount>;
    let launcher: MySQLShellLauncher;

    const started = (): Promise<void> => {
        return new Promise((resolve) => {
            const loaded = (): Promise<boolean> => {
                requisitions.unregister("applicationDidStart", loaded);
                resolve();

                return Promise.resolve(true);
            };

            requisitions.register("applicationDidStart", loaded);
        });
    };

    beforeAll(async () => {
        // No automatic login takes place here, like in other tests. The app will trigger the login.
        launcher = await setupShellForTests("Application", false, false);
        expect(MessageScheduler.get.isConnected).toBe(false);

        expect(requisitions.registrations("dialogResponse")).toBe(1);

        app = mount(
            <App />,
        );
        await started();

        expect(requisitions.registrations("dialogResponse")).toBe(2);
    });

    afterAll(async () => {
        const event = new Event("beforeunload");
        expect(window.dispatchEvent(event)).toBe(true);

        MessageScheduler.get.disconnect();
        expect(MessageScheduler.get.isConnected).toBe(false);

        app.unmount();

        await launcher.exitProcess();
    });

    it("Application readiness", () => {
        // Note: at this point the app should be fully up and debugging the code seems to show that the
        //       application host is rendered. Yet the snapshot still shows the login page.
        //       What's even more confusing is that we come here, because the application host sent a message
        //       that it was mounted. Need to investigate later.
        expect(app.state("loginInProgress")).toEqual(false);

        expect(snapshotFromWrapper(app)).toMatchSnapshot();
    });

    it("Fake error event", async () => {
        await dispatchErrorResponse();
    });

    it("Handling status bar click events", async () => {
        let result = await requisitions.execute("statusBarButtonClick", { type: "dummy", event: eventMock });
        expect(result).toBe(false);

        result = await requisitions.execute("statusBarButtonClick", { type: "openPopupMenu", event: eventMock });
        expect(result).toBe(true);
    });

    it("Handling status bar update events", async () => {
        let lastItems: IStatusbarInfo[] = [];
        const listener = (items: IStatusbarInfo[]): Promise<boolean> => {
            lastItems = items;

            return Promise.resolve(true);
        };

        requisitions.register("updateStatusbar", listener);

        const statusInfo: IEditorStatusInfo = {};

        let result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([]);

        statusInfo.indentSize = 4;
        statusInfo.tabSize = 4;
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorIndent",
            visible: true,
            text: "Tab Size: 4",
        }]);

        statusInfo.insertSpaces = true;
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorIndent",
            visible: true,
            text: "Spaces: 4",
        }]);

        delete statusInfo.indentSize;
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorIndent",
            visible: true,
            text: "Spaces: 0",
        }]);

        statusInfo.indentSize = -1;
        delete statusInfo.tabSize;
        statusInfo.insertSpaces = false;
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorIndent",
            visible: true,
            text: "Tab Size: 0",
        }]);

        statusInfo.line = 22;
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorIndent",
            visible: true,
            text: "Tab Size: 0",
        },
        {
            id: "editorPosition",
            visible: true,
            text: "Ln 22, Col 1",
        }]);

        delete statusInfo.indentSize;
        statusInfo.column = 33;
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorPosition",
            visible: true,
            text: "Ln 22, Col 33",
        }]);

        delete statusInfo.line;
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorPosition",
            visible: true,
            text: "Ln 1, Col 33",
        }]);

        statusInfo.language = "JavaScript";
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorPosition",
            visible: true,
            text: "Ln 1, Col 33",
        },
        {
            id: "editorLanguage",
            visible: true,
            text: "JavaScript",
        }]);

        statusInfo.eol = "LF";
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorPosition",
            visible: true,
            text: "Ln 1, Col 33",
        },
        {
            id: "editorLanguage",
            visible: true,
            text: "JavaScript",
        },
        {
            id: "editorEOL",
            visible: true,
            text: "LF",
        }]);

        statusInfo.indentSize = 4;
        statusInfo.tabSize = 4;
        result = await requisitions.execute("editorInfoUpdated", statusInfo);
        expect(result).toBe(true);
        expect(lastItems).toEqual([{
            id: "editorIndent",
            visible: true,
            text: "Tab Size: 4",
        },
        {
            id: "editorPosition",
            visible: true,
            text: "Ln 1, Col 33",
        },
        {
            id: "editorLanguage",
            visible: true,
            text: "JavaScript",
        },
        {
            id: "editorEOL",
            visible: true,
            text: "LF",
        }]);

        requisitions.unregister("updateStatusbar", listener);
    });

    it("DOM events", () => {
        const event = new MouseEvent("contextmenu", { cancelable: true });
        expect(event.defaultPrevented).toBe(false);
        expect(window.document.body.dispatchEvent(event)).toBe(false); // It's cancelled in App.tsx.
        expect(event.defaultPrevented).toBe(true);
    });

    it("Other events", async () => {
        let messageSent = false;

        // Pretend we are embedded.
        appParameters.embedded = true;
        const listener = (event: MessageEvent): void => {
            if (event.data.command === "themeChanged" && event.data.source === "app") {
                if (event.data.data.name === "My Theme" && event.data.data.type === "dark") {
                    messageSent = true;
                }
            }
        };

        window.addEventListener("message", listener);
        const listenerPromise = waitFor(1000, () => {
            return messageSent;
        });

        const requisitionPromise = await requisitions.execute("themeChanged", {
            name: "My Theme",
            type: "dark",
            values: {
                name: "Color 1",
            },
        });
        const result = await Promise.all([listenerPromise, requisitionPromise]);
        expect(result.length).toBe(2);
        expect(result[0]).toBe(true);
        expect(result[1]).toBe(true);

        window.removeEventListener("message", listener);
    });
});
