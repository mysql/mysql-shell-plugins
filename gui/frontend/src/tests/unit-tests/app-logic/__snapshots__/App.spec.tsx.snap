// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application tests Application readiness 1`] = `
<App>
  <ErrorBoundary>
    <LoginPage>
      <Container
        id="loginDialog"
        orientation="column"
      >
        <div
          className="msg container fixedScrollbar"
          id="loginDialog"
          style={
            Object {
              "alignItems": undefined,
              "flexDirection": "column",
              "flexWrap": undefined,
              "justifyContent": undefined,
            }
          }
        >
          <Icon
            disabled={false}
            id="loginDialogSakilaLogo"
            src="module-shell.svg"
          >
            <div
              className="msg icon"
              id="loginDialogSakilaLogo"
              style={
                Object {
                  "WebkitMaskImage": "url(module-shell.svg)",
                  "backgroundColor": undefined,
                  "height": undefined,
                  "maskImage": "url(module-shell.svg)",
                  "width": undefined,
                }
              }
            />
          </Icon>
          <Container
            id="heading"
            orientation="column"
          >
            <div
              className="msg container fixedScrollbar"
              id="heading"
              style={
                Object {
                  "alignItems": undefined,
                  "flexDirection": "column",
                  "flexWrap": undefined,
                  "justifyContent": undefined,
                }
              }
            >
              <Label
                as="h2"
                id="headingLabel"
              >
                <h2
                  className="msg label"
                  data-tooltip="expand"
                  id="headingLabel"
                  style={
                    Object {
                      "textAlign": undefined,
                    }
                  }
                >
                  MySQL Shell
                </h2>
              </Label>
              <Label
                id="headingSubLabel"
              >
                <label
                  className="msg label"
                  data-tooltip="expand"
                  id="headingSubLabel"
                  style={
                    Object {
                      "textAlign": undefined,
                    }
                  }
                >
                  Welcome to the MySQL Shell GUI.
                  <br />
                  Please provide your MySQL Shell GUI credentials to log into the shell interface.
                </label>
              </Label>
            </div>
          </Container>
          <Container
            id="loginDialogLinks"
            mainAlignment="center"
            orientation="row"
            wrap="wrap"
          >
            <div
              className="msg container fixedScrollbar"
              id="loginDialogLinks"
              style={
                Object {
                  "alignItems": undefined,
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "center",
                }
              }
            >
              <a
                href="http://localhost:3001/#"
                key="Learn More >"
              >
                Learn More &gt;
              </a>
              <a
                href="http://localhost:3001/#"
                key="Browse Tutorial >"
              >
                Browse Tutorial &gt;
              </a>
              <a
                href="http://localhost:3001/#"
                key="Read Docs >"
              >
                Read Docs &gt;
              </a>
            </div>
          </Container>
          <Grid
            columnGap={8}
            columns={
              Array [
                220,
                40,
              ]
            }
            id="loginDialogControls"
          >
            <div
              className="msg grid"
              id="loginDialogControls"
              style={
                Object {
                  "columnGap": "8px",
                  "gridTemplateColumns": "220px 40px",
                  "rowGap": undefined,
                }
              }
            >
              <GridCell
                orientation="column"
              >
                <Container
                  className="msg gridCell"
                  orientation="column"
                  style={Object {}}
                >
                  <div
                    className="msg container fixedScrollbar gridCell"
                    style={
                      Object {
                        "alignItems": undefined,
                        "flexDirection": "column",
                        "flexWrap": undefined,
                        "justifyContent": undefined,
                      }
                    }
                  >
                    <Input
                      id="loginUsername"
                      onChange={[Function]}
                      onConfirm={[Function]}
                      placeholder="Username"
                      spellCheck={false}
                      value=""
                    >
                      <input
                        className="msg input"
                        id="loginUsername"
                        onKeyDown={[Function]}
                        oninput={[Function]}
                        placeholder="Username"
                        spellCheck={false}
                        style={
                          Object {
                            "textAlign": undefined,
                          }
                        }
                        type="text"
                        value=""
                      />
                    </Input>
                  </div>
                </Container>
              </GridCell>
              <GridCell>
                <Container
                  className="msg gridCell"
                  orientation="row"
                  style={Object {}}
                >
                  <div
                    className="msg container fixedScrollbar gridCell"
                    style={
                      Object {
                        "alignItems": undefined,
                        "flexDirection": "row",
                        "flexWrap": undefined,
                        "justifyContent": undefined,
                      }
                    }
                  />
                </Container>
              </GridCell>
              <GridCell
                orientation="column"
              >
                <Container
                  className="msg gridCell"
                  orientation="column"
                  style={Object {}}
                >
                  <div
                    className="msg container fixedScrollbar gridCell"
                    style={
                      Object {
                        "alignItems": undefined,
                        "flexDirection": "column",
                        "flexWrap": undefined,
                        "justifyContent": undefined,
                      }
                    }
                  >
                    <Input
                      id="loginPassword"
                      onChange={[Function]}
                      onConfirm={[Function]}
                      password={true}
                      placeholder="Password"
                      spellCheck={true}
                      value=""
                    >
                      <input
                        className="msg input"
                        id="loginPassword"
                        onKeyDown={[Function]}
                        oninput={[Function]}
                        placeholder="Password"
                        spellCheck={true}
                        style={
                          Object {
                            "textAlign": undefined,
                          }
                        }
                        type="password"
                        value=""
                      />
                    </Input>
                  </div>
                </Container>
              </GridCell>
              <GridCell
                mainAlignment="center"
                orientation="column"
              >
                <Container
                  className="msg gridCell"
                  mainAlignment="center"
                  orientation="column"
                  style={Object {}}
                >
                  <div
                    className="msg container fixedScrollbar gridCell"
                    style={
                      Object {
                        "alignItems": undefined,
                        "flexDirection": "column",
                        "flexWrap": undefined,
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Button
                      id="loginButton"
                      onClick={[Function]}
                      round={true}
                    >
                      <button
                        className="msg button round"
                        id="loginButton"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        style={
                          Object {
                            "flexDirection": undefined,
                          }
                        }
                        tabIndex={0}
                      >
                        <Icon
                          disabled={false}
                          src="chevron-right.svg"
                        >
                          <div
                            className="msg icon"
                            style={
                              Object {
                                "WebkitMaskImage": "url(chevron-right.svg)",
                                "backgroundColor": undefined,
                                "height": undefined,
                                "maskImage": "url(chevron-right.svg)",
                                "width": undefined,
                              }
                            }
                          />
                        </Icon>
                      </button>
                    </Button>
                  </div>
                </Container>
              </GridCell>
            </div>
          </Grid>
          <Label
            id="loginDialogSubtext"
          >
            <label
              className="msg label"
              data-tooltip="expand"
              id="loginDialogSubtext"
              style={
                Object {
                  "textAlign": undefined,
                }
              }
            >
              If you do not have an MySQL Shell GUI account yet, please
              <br />
              ask your administrator to have one created for you.
            </label>
          </Label>
          <div
            className="copyright"
          />
        </div>
      </Container>
    </LoginPage>
    <ErrorPanel>
      <Dialog
        actions={
          Object {
            "end": Array [
              <Button
                caption="Close"
                onClick={[Function]}
              />,
            ],
          }
        }
        caption={
          <y>
            <Icon
              disabled={false}
              src={123}
            />
            <Label>
              
            </Label>
          </y>
        }
        className="msg errorPanel"
        container={Object {}}
        content={
          <Container
            orientation="column"
          >
            
          </Container>
        }
      >
        <Portal
          container={Object {}}
          onClose={[Function]}
          onOpen={[Function]}
        />
      </Dialog>
    </ErrorPanel>
    <TooltipProvider
      showDelay={200}
    />
    <Statusbar
      items={
        Array [
          Object {
            "id": "profileTitle",
            "text": "Profile:",
            "type": 0,
            "visible": false,
          },
          Object {
            "choices": Array [],
            "commandId": "openPopupMenu",
            "id": "profileChoice",
            "tooltip": "Change profile",
            "type": 1,
            "visible": false,
          },
          Object {
            "id": "details",
            "rightAlign": true,
            "standout": false,
            "type": 0,
          },
          Object {
            "id": "editorLanguage",
            "rightAlign": true,
            "type": 0,
            "visible": false,
          },
          Object {
            "id": "editorEOL",
            "rightAlign": true,
            "type": 0,
            "visible": false,
          },
          Object {
            "id": "editorIndent",
            "rightAlign": true,
            "type": 0,
            "visible": false,
          },
          Object {
            "commandId": "scripting:editor.action.gotoLine",
            "id": "editorPosition",
            "rightAlign": true,
            "type": 1,
            "visible": false,
          },
        ]
      }
    >
      <Container
        className="msg statusbar verticalCenterContent"
        orientation="row"
      >
        <div
          className="msg container fixedScrollbar statusbar verticalCenterContent"
          style={
            Object {
              "alignItems": undefined,
              "flexDirection": "row",
              "flexWrap": undefined,
              "justifyContent": undefined,
            }
          }
        >
          <Container
            className="leftItems"
            crossAlignment="stretch"
            orientation="row"
          >
            <div
              className="msg container fixedScrollbar leftItems"
              style={
                Object {
                  "alignItems": "stretch",
                  "flexDirection": "row",
                  "flexWrap": undefined,
                  "justifyContent": undefined,
                }
              }
            />
          </Container>
          <Container
            className="rightItems"
            crossAlignment="stretch"
            orientation="row-reverse"
          >
            <div
              className="msg container fixedScrollbar rightItems"
              style={
                Object {
                  "alignItems": "stretch",
                  "flexDirection": "row-reverse",
                  "flexWrap": undefined,
                  "justifyContent": undefined,
                }
              }
            />
          </Container>
        </div>
      </Container>
    </Statusbar>
    <ProfileSelector>
      <ValueEditDialog
        id="profileCreating"
        onClose={[Function]}
        onValidate={[Function]}
      >
        <Dialog
          actions={
            Object {
              "begin": Array [],
              "end": Array [
                <Button
                  caption="OK"
                  disabled={false}
                  id="ok"
                  onClick={[Function]}
                />,
                <Button
                  caption="Cancel"
                  id="cancel"
                  onClick={[Function]}
                />,
              ],
            }
          }
          caption={
            <y>
              <Icon
                disabled={false}
                src={388}
              />
              <Label />
            </y>
          }
          className="msg valueEditDialog"
          container={Object {}}
          content={
            <Grid
              columnGap={16}
              columns={8}
              rowGap={16}
            />
          }
          id="profileCreating"
          onClose={[Function]}
        >
          <Portal
            container={Object {}}
            id="profileCreating"
            onClose={[Function]}
            onOpen={[Function]}
          />
        </Dialog>
      </ValueEditDialog>
      <ConfirmDialog
        id="confirmDeleteDialog"
        onClose={[Function]}
      >
        <Dialog
          actions={
            Object {
              "end": Array [],
            }
          }
          caption={
            <y>
              <Icon
                disabled={false}
                src={275}
              />
              <Label>
                Confirm
              </Label>
            </y>
          }
          className="msg confirmDialog"
          container={Object {}}
          content={
            <Container
              orientation="column"
            >
              
              <Container
                className="description"
                orientation="column"
              />
            </Container>
          }
          onClose={[Function]}
        >
          <Portal
            container={Object {}}
            onClose={[Function]}
            onOpen={[Function]}
          />
        </Dialog>
      </ConfirmDialog>
      <Menu
        id="tileActionMenu"
        onItemClick={[Function]}
        orientation="column"
        placement="top-start"
      >
        <Popup
          className="msg menu"
          id="tileActionMenu"
          onClose={[Function]}
          onKeyDown={[Function]}
          onOpen={[Function]}
          orientation="column"
          pinned={false}
          placement="top-start"
          showArrow={false}
        >
          <Portal
            id="tileActionMenu"
            onClose={[Function]}
            onKeyDown={[Function]}
            onOpen={[Function]}
          />
        </Popup>
      </Menu>
    </ProfileSelector>
    <ColorPopup>
      <Popup
        className="msg colorPopup"
        onOpen={[Function]}
        orientation="column"
        pinned={false}
        placement="bottom"
        showArrow={true}
      >
        <Portal
          onClose={[Function]}
          onOpen={[Function]}
        />
      </Popup>
    </ColorPopup>
  </ErrorBoundary>
</App>
`;
